src/node_crypto.cc

```diff
@@ -1972,6 +1985,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
 }
 #endif  // OPENSSL_NPN_NEGOTIATED
 
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+typedef struct tlsextalpnctx_st {
+  unsigned char *data;
+  unsigned short len;
+} tlsextalpnctx;
+
+template <class Base>
+int SSLWrap<Base>::SelectALPNCallback(SSL* s,
+                                      const unsigned char** out,
+                                      unsigned char* outlen,
+                                      const unsigned char* in,
+                                      unsigned int inlen,
+                                      void* arg) {
+  Base* w = static_cast<Base*>(SSL_get_app_data(s));
+  Environment* env = w->env();
+  HandleScope handle_scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  int status = SSL_select_next_proto((unsigned char**) out, outlen,
+                                     w->alpn_protos_, w->alpn_protos_len_,
+                                     in, inlen);
+
+  switch (status) {
+    case OPENSSL_NPN_NO_OVERLAP:
+      // According to 3.2. Protocol Selection of RFC7301,
+      // fatal no_application_protocol alert shall be sent
+      // but current openssl does not support it yet.
+      // See https://rt.openssl.org/Ticket/Display.html?id=3463.
+      // Instead, we send a warning alert for now.
+      return SSL_TLSEXT_ERR_ALERT_WARNING;
+    case OPENSSL_NPN_NEGOTIATED:
+      return SSL_TLSEXT_ERR_OK;
+    default:
+      return SSL_TLSEXT_ERR_ALERT_FATAL;
+  }
+}
+#endif  // TLSEXT_TYPE_application_layer_protocol_negotiation
+
+
+template <class Base>
+void SSLWrap<Base>::GetALPNNegotiatedProto(
+    const FunctionCallbackInfo<v8::Value>& args) {
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+  HandleScope scope(args.GetIsolate());
+  Base* w = Unwrap<Base>(args.Holder());
+
+  const unsigned char* alpn_proto;
+  unsigned int alpn_proto_len;
+
+  SSL_get0_alpn_selected(w->ssl_, &alpn_proto, &alpn_proto_len);
+
+  if (!alpn_proto)
+    return args.GetReturnValue().Set(false);
+
+  args.GetReturnValue().Set(
+      OneByteString(args.GetIsolate(), alpn_proto, alpn_proto_len));
+#endif  // TLSEXT_TYPE_application_layer_protocol_negotiation
+}
+
+
+template <class Base>
+void SSLWrap<Base>::SetALPNProtocols(
+    const FunctionCallbackInfo<v8::Value>& args) {
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+  HandleScope scope(args.GetIsolate());
+  Base* w = Unwrap<Base>(args.Holder());
+  if (args.Length() < 1 || !Buffer::HasInstance(args[0]))
+    return w->env()->ThrowTypeError("Must give a Buffer as first argument");
+
+    if (w->is_client()) {
+      const unsigned char* alpn_protos =
+          reinterpret_cast<const unsigned char*>(Buffer::Data(args[0]));
+      unsigned alpn_protos_len = Buffer::Length(args[0]);
+      int r = SSL_set_alpn_protos(w->ssl_, alpn_protos, alpn_protos_len);
+      assert(r == 0);
+    } else {
+      w->alpn_protos_ =
```

Comment:
> What do you think about allocating and copying the data instead? I think I'm mostly against using Persistent for this reasons now.

Let me confirm this. Is this what you mean or another way to avoid using Persistent? Maybe I lost the point. Thanks.
```diff
@@ -2063,7 +2063,10 @@ void SSLWrap<Base>::SetALPNProtocols(
       int r = SSL_set_alpn_protos(w->ssl_, alpn_protos, alpn_protos_len);
       assert(r == 0);
     } else {
-      w->alpn_protos_.Reset(args.GetIsolate(), args[0].As<Object>());
+      w->alpn_protos_.Reset(args.GetIsolate(), Buffer::Copy(
+          w->env(),
+          reinterpret_cast<const char*>(Buffer::Data(args[0])),
+          Buffer::Length(args[0])).ToLocalChecked());
       // Server should select ALPN protocol from list of advertised by client
       SSL_CTX_set_alpn_select_cb(w->ssl_->ctx, SelectALPNCallback, nullptr);
     }
````